plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/exporter/QueryInsightsLocalIndexExporter.java

```diff
@@ -0,0 +1,216 @@
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Class to export data collected by search query analyzers to a local OpenSearch index
+ * <p>
+ * Internal used within the Query Insight framework
+ *
+ * @opensearch.internal
+ */
+public class QueryInsightsLocalIndexExporter<T extends SearchQueryRecord<?>> extends QueryInsightsExporter<T> {
+    private static final Logger log = LogManager.getLogger(QueryInsightsLocalIndexExporter.class);
+    private static final int INDEX_TIMEOUT = 60;
+
+    private final ClusterService clusterService;
+    private final Client client;
+
+    /** The mapping for the local index that holds the data */
+    private final InputStream localIndexMapping;
+
+    public QueryInsightsLocalIndexExporter(
+        ClusterService clusterService,
+        Client client,
+        String localIndexName,
+        InputStream localIndexMapping
+    ) {
+        super(QueryInsightsExporterType.LOCAL_INDEX, localIndexName);
+        this.clusterService = clusterService;
+        this.client = client;
+        this.localIndexMapping = localIndexMapping;
+    }
+
+    /**
+     * Export the data to the predefined local OpenSearch Index
+     *
+     * @param records the data to export
+     * @throws IOException if an error occurs
+     */
+    @Override
+    public synchronized void export(List<T> records) throws IOException {
+        if (records.size() == 0) {
+            return;
+        }
+        if (checkIfIndexExists()) {
+            bulkRecord(records);
+        } else {
+            // local index not exist
+            initLocalIndex(new ActionListener<>() {
+                @Override
+                public void onResponse(CreateIndexResponse response) {
+                    if (response.isAcknowledged()) {
+                        log.debug(
+                            String.format(Locale.ROOT, "successfully initialized local index %s for query insight.", getIdentifier())
+                        );
+                        try {
+                            bulkRecord(records);
+                        } catch (IOException e) {
+                            log.error(String.format(Locale.ROOT, "fail to ingest query insight data to local index, error: %s", e));
+                        }
+                    } else {
+                        log.error(
+                            String.format(
+                                Locale.ROOT,
+                                "request to created local index %s for query insight not acknowledged.",
+                                getIdentifier()
+                            )
+                        );
+                    }
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    log.error(String.format(Locale.ROOT, "error creating local index for query insight: %s", e));
+                }
```

Comment:
Should the export method silently log the error and complete? It should let the caller know it failed, also should there be some async attempts?

Reply:
The caller of this function is internal services. Other than logging the error, we can probably emit some metrics in the future to indicate this error. But yeah I think it make sense to retry in some cases.